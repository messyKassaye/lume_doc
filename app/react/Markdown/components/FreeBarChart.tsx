/** @format */

import React, { Component } from 'react';

import {
  ResponsiveContainer,
  BarChart,
  XAxis,
  YAxis,
  CartesianGrid,
  Bar,
  Tooltip,
  Cell,
} from 'recharts';

type layoutType = 'horizontal' | 'vertical' | undefined;

type XProps = {
  layout: layoutType;
  dataKey: string;
};

type YProps = { layout: layoutType };

const X = ({ layout, dataKey }: XProps) =>
  layout === 'vertical' ? (
    <XAxis type="number" dataKey={dataKey} />
  ) : (
    <XAxis dataKey="label" label="" />
  );

const Y = ({ layout }: YProps) =>
  layout === 'vertical' ? <YAxis width={200} type="category" dataKey="label" /> : <YAxis />;

export type CustomTooltipProps = {
  active: boolean;
  payload?: any[];
  label?: string;
};

export const CustomTooltip = ({ active, payload, label }: CustomTooltipProps) => {
  if (active && payload && payload[0]) {
    const finalItemStyle = {
      display: 'block',
      color: '#999',
      backgroundColor: 'rgba(255,255,255,0.9)',
      border: '1px solid #ccc',
      padding: '10px',
    };
    return (
      <div className="custom-tooltip" style={finalItemStyle}>
        <p className="tooltip-label" style={{ margin: 0 }}>
          {label} :{' '}
          <span className="tooltip-label-value" style={{ color: payload[0].color || '#000' }}>
            {payload[0].value}
          </span>
        </p>
      </div>
    );
  }

  return null;
};

CustomTooltip.defaultProps = {
  active: false,
};

export type FreeBarChartProps = {
  classname: string;
  layout: layoutType;
  data: string | null;
  dataKeys: string;
  colors: string;
  children: any;
};

type FreeBarChartState = {
  activeDataIndex: number;
};

class FreeBarChart extends Component<FreeBarChartProps, FreeBarChartState> {
  static defaultProps: Partial<FreeBarChartProps>;

  constructor(props: FreeBarChartProps) {
    super(props);
    this.state = { activeDataIndex: 0 };
    this.changeData = this.changeData.bind(this);
  }

  getDataKeys() {
    const { dataKeys } = this.props;
    if (dataKeys.indexOf('[') === 0) {
      return JSON.parse(dataKeys);
    }

    return [{ [dataKeys]: dataKeys }];
  }

  changeData(activeDataIndex: number) {
    return () => {
      this.setState({ activeDataIndex });
    };
  }

  render() {
    const { activeDataIndex } = this.state;
    const { layout, data, classname, colors, children } = this.props;
    let output = null;

    if (data) {
      const sliceColors = colors.split(',');
      const parsedData = JSON.parse(data);
      const dataKeysArray = this.getDataKeys();
      const dataKey = Object.keys(dataKeysArray[activeDataIndex])[0];
      output = (
        <>
          {dataKeysArray.length > 1 && (
            <div className="toggle-group">
              {dataKeysArray.map((dataKeyValue: object, index: number) => (
                <button
                  type="button"
                  className={`btn ${
                    activeDataIndex === index ? 'btn-primary' : 'btn-default'
                  } toggle-group-button`}
                  key={Object.keys(dataKeyValue)[0]}
                  onClick={this.changeData(index)}
                >
                  {Object.values(dataKeyValue)[0]}
                </button>
              ))}
            </div>
          )}
          {children}
          <ResponsiveContainer height={320}>
            <BarChart height={300} data={parsedData} layout={layout}>
              {X({ layout, dataKey })}
              {Y({ layout })}
              <CartesianGrid strokeDasharray="2 4" />
              <Tooltip content={<CustomTooltip />} />
              <Bar dataKey={dataKey} fill="rgb(30, 28, 138)" stackId="unique">
                {parsedData.map((_entry: any, index: number) => (
                  <Cell
                    cursor="pointer"
                    fill={sliceColors[index % sliceColors.length]}
                    key={`cell-${index}`} // eslint-disable-line react/no-array-index-key
                  />
                ))}
              </Bar>
            </BarChart>
          </ResponsiveContainer>
        </>
      );
    }

    return <div className={`BarChart ${classname}`}>{output}</div>;
  }
}

FreeBarChart.defaultProps = {
  classname: '',
  layout: 'horizontal',
  data: null,
  dataKeys: 'results',
  colors: '#1e1c8a',
  children: null,
};

export default FreeBarChart;
