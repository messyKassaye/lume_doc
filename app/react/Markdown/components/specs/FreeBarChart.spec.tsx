/**
 * @format
 * @jest-environment jsdom
 */

import React from 'react';
import { shallow } from 'enzyme';

import FreeBarChart, {
  FreeBarChartProps,
  CustomTooltip,
  CustomTooltipProps,
} from '../FreeBarChart';

describe('FreeBarChart', () => {
  let component: any;

  let props: Partial<FreeBarChartProps>;

  const render = () => {
    component = shallow<FreeBarChart>(<FreeBarChart {...props} />);
  };

  const expectMatch = () => {
    render();
    expect(component).toMatchSnapshot();
  };

  it('should output empty div if no data', () => {
    expectMatch();
  });

  describe('when data', () => {
    beforeEach(() => {
      props = {
        data:
          '[{"label":"col1", "results": 3, "otherProp": 5, "lastProp": 2},{"label":"col2", "results": 4, "otherProp": 2, "lastProp": 3}]',
      };
    });

    it('should render basic data with correct defaults', () => {
      expectMatch();
    });

    it('should allow personalizing the options', () => {
      props.classname = 'custom-class';
      props.layout = 'vertical';
      props.dataKeys = '[{ "otherProp": "Button label" }]';
      props.colors = '#ff0,#00f';
      expectMatch();
    });

    it('should allow changing dataKeys with buttons', () => {
      props.dataKeys = '[{ "otherProp": "Button 1 label" }, { "lastProp": "Button 2 label" }]';
      expectMatch();
      const lastPropButton = component.find('button').at(1);
      lastPropButton.simulate('click');
      expect(component).toMatchSnapshot();
    });
  });
});

describe('CustomTooltip', () => {
  let props: Partial<CustomTooltipProps>;
  let component: any;

  const render = () => {
    component = shallow(<CustomTooltip {...props} />);
  };

  const expectMatch = () => {
    render();
    expect(component).toMatchSnapshot();
  };

  it('should not render on inactive', () => {
    expectMatch();
  });

  it('should render the tooltip on active', () => {
    props = { active: true, payload: [{ value: 3 }], label: 'tooltipLabel' };
    expectMatch();
  });

  it('should render with payload color', () => {
    props = { active: true, payload: [{ value: 5, color: '#f00' }], label: 'tooltipLabel' };
    expectMatch();
  });
});
