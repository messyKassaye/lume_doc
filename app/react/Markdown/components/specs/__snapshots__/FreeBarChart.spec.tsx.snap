// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomTooltip should not render on inactive 1`] = `""`;

exports[`CustomTooltip should render the tooltip on active 1`] = `
<div
  className="custom-tooltip"
  style={
    Object {
      "backgroundColor": "rgba(255,255,255,0.9)",
      "border": "1px solid #ccc",
      "color": "#999",
      "display": "block",
      "padding": "10px",
    }
  }
>
  <p
    className="tooltip-label"
    style={
      Object {
        "margin": 0,
      }
    }
  >
    tooltipLabel
     :
     
    <span
      className="tooltip-label-value"
      style={
        Object {
          "color": "#000",
        }
      }
    >
      3
    </span>
  </p>
</div>
`;

exports[`CustomTooltip should render with payload color 1`] = `
<div
  className="custom-tooltip"
  style={
    Object {
      "backgroundColor": "rgba(255,255,255,0.9)",
      "border": "1px solid #ccc",
      "color": "#999",
      "display": "block",
      "padding": "10px",
    }
  }
>
  <p
    className="tooltip-label"
    style={
      Object {
        "margin": 0,
      }
    }
  >
    tooltipLabel
     :
     
    <span
      className="tooltip-label-value"
      style={
        Object {
          "color": "#f00",
        }
      }
    >
      5
    </span>
  </p>
</div>
`;

exports[`FreeBarChart should output empty div if no data 1`] = `
<div
  className="BarChart "
/>
`;

exports[`FreeBarChart when data should allow changing dataKeys with buttons 1`] = `
<div
  className="BarChart "
>
  <div
    className="toggle-group"
  >
    <button
      className="btn btn-primary toggle-group-button"
      key="otherProp"
      onClick={[Function]}
      type="button"
    >
      Button 1 label
    </button>
    <button
      className="btn btn-default toggle-group-button"
      key="lastProp"
      onClick={[Function]}
      type="button"
    >
      Button 2 label
    </button>
  </div>
  <ResponsiveContainer
    debounce={0}
    height={320}
    width="100%"
  >
    <BarChart
      barCategoryGap="10%"
      barGap={4}
      data={
        Array [
          Object {
            "label": "col1",
            "lastProp": 2,
            "otherProp": 5,
            "results": 3,
          },
          Object {
            "label": "col2",
            "lastProp": 3,
            "otherProp": 2,
            "results": 4,
          },
        ]
      }
      height={300}
      layout="horizontal"
      margin={
        Object {
          "bottom": 5,
          "left": 5,
          "right": 5,
          "top": 5,
        }
      }
      reverseStackOrder={false}
      stackOffset="none"
    >
      <XAxis
        allowDataOverflow={false}
        allowDecimals={true}
        allowDuplicatedCategory={true}
        dataKey="label"
        domain={
          Array [
            0,
            "auto",
          ]
        }
        height={30}
        hide={false}
        label=""
        mirror={false}
        orientation="bottom"
        padding={
          Object {
            "left": 0,
            "right": 0,
          }
        }
        reversed={false}
        scale="auto"
        tickCount={5}
        type="category"
        width={0}
        xAxisId={0}
      />
      <YAxis
        allowDataOverflow={false}
        allowDecimals={true}
        allowDuplicatedCategory={true}
        domain={
          Array [
            0,
            "auto",
          ]
        }
        height={0}
        hide={false}
        mirror={false}
        orientation="left"
        padding={
          Object {
            "bottom": 0,
            "top": 0,
          }
        }
        reversed={false}
        scale="auto"
        tickCount={5}
        type="number"
        width={60}
        yAxisId={0}
      />
      <CartesianGrid
        fill="none"
        horizontal={true}
        horizontalFill={Array []}
        horizontalPoints={Array []}
        stroke="#ccc"
        strokeDasharray="2 4"
        vertical={true}
        verticalFill={Array []}
        verticalPoints={Array []}
      />
      <Tooltip
        active={false}
        animationDuration={400}
        animationEasing="ease"
        content={
          <CustomTooltip
            active={false}
          />
        }
        contentStyle={Object {}}
        coordinate={
          Object {
            "x": 0,
            "y": 0,
          }
        }
        cursor={true}
        cursorStyle={Object {}}
        filterNull={true}
        isAnimationActive={true}
        itemSorter={[Function]}
        itemStyle={Object {}}
        labelStyle={Object {}}
        offset={10}
        separator=" : "
        useTranslate3d={false}
        viewBox={
          Object {
            "x1": 0,
            "x2": 0,
            "y1": 0,
            "y2": 0,
          }
        }
        wrapperStyle={Object {}}
      />
      <Bar
        animationBegin={0}
        animationDuration={400}
        animationEasing="ease"
        data={Array []}
        dataKey="otherProp"
        fill="rgb(30, 28, 138)"
        hide={false}
        isAnimationActive={true}
        layout="vertical"
        legendType="rect"
        minPointSize={0}
        onAnimationEnd={[Function]}
        onAnimationStart={[Function]}
        stackId="unique"
        xAxisId={0}
        yAxisId={0}
      >
        <Cell
          cursor="pointer"
          fill="#1e1c8a"
          key="cell-0"
        />
        <Cell
          cursor="pointer"
          fill="#1e1c8a"
          key="cell-1"
        />
      </Bar>
    </BarChart>
  </ResponsiveContainer>
</div>
`;

exports[`FreeBarChart when data should allow changing dataKeys with buttons 2`] = `
<div
  className="BarChart "
>
  <div
    className="toggle-group"
  >
    <button
      className="btn btn-default toggle-group-button"
      key="otherProp"
      onClick={[Function]}
      type="button"
    >
      Button 1 label
    </button>
    <button
      className="btn btn-primary toggle-group-button"
      key="lastProp"
      onClick={[Function]}
      type="button"
    >
      Button 2 label
    </button>
  </div>
  <ResponsiveContainer
    debounce={0}
    height={320}
    width="100%"
  >
    <BarChart
      barCategoryGap="10%"
      barGap={4}
      data={
        Array [
          Object {
            "label": "col1",
            "lastProp": 2,
            "otherProp": 5,
            "results": 3,
          },
          Object {
            "label": "col2",
            "lastProp": 3,
            "otherProp": 2,
            "results": 4,
          },
        ]
      }
      height={300}
      layout="horizontal"
      margin={
        Object {
          "bottom": 5,
          "left": 5,
          "right": 5,
          "top": 5,
        }
      }
      reverseStackOrder={false}
      stackOffset="none"
    >
      <XAxis
        allowDataOverflow={false}
        allowDecimals={true}
        allowDuplicatedCategory={true}
        dataKey="label"
        domain={
          Array [
            0,
            "auto",
          ]
        }
        height={30}
        hide={false}
        label=""
        mirror={false}
        orientation="bottom"
        padding={
          Object {
            "left": 0,
            "right": 0,
          }
        }
        reversed={false}
        scale="auto"
        tickCount={5}
        type="category"
        width={0}
        xAxisId={0}
      />
      <YAxis
        allowDataOverflow={false}
        allowDecimals={true}
        allowDuplicatedCategory={true}
        domain={
          Array [
            0,
            "auto",
          ]
        }
        height={0}
        hide={false}
        mirror={false}
        orientation="left"
        padding={
          Object {
            "bottom": 0,
            "top": 0,
          }
        }
        reversed={false}
        scale="auto"
        tickCount={5}
        type="number"
        width={60}
        yAxisId={0}
      />
      <CartesianGrid
        fill="none"
        horizontal={true}
        horizontalFill={Array []}
        horizontalPoints={Array []}
        stroke="#ccc"
        strokeDasharray="2 4"
        vertical={true}
        verticalFill={Array []}
        verticalPoints={Array []}
      />
      <Tooltip
        active={false}
        animationDuration={400}
        animationEasing="ease"
        content={
          <CustomTooltip
            active={false}
          />
        }
        contentStyle={Object {}}
        coordinate={
          Object {
            "x": 0,
            "y": 0,
          }
        }
        cursor={true}
        cursorStyle={Object {}}
        filterNull={true}
        isAnimationActive={true}
        itemSorter={[Function]}
        itemStyle={Object {}}
        labelStyle={Object {}}
        offset={10}
        separator=" : "
        useTranslate3d={false}
        viewBox={
          Object {
            "x1": 0,
            "x2": 0,
            "y1": 0,
            "y2": 0,
          }
        }
        wrapperStyle={Object {}}
      />
      <Bar
        animationBegin={0}
        animationDuration={400}
        animationEasing="ease"
        data={Array []}
        dataKey="lastProp"
        fill="rgb(30, 28, 138)"
        hide={false}
        isAnimationActive={true}
        layout="vertical"
        legendType="rect"
        minPointSize={0}
        onAnimationEnd={[Function]}
        onAnimationStart={[Function]}
        stackId="unique"
        xAxisId={0}
        yAxisId={0}
      >
        <Cell
          cursor="pointer"
          fill="#1e1c8a"
          key="cell-0"
        />
        <Cell
          cursor="pointer"
          fill="#1e1c8a"
          key="cell-1"
        />
      </Bar>
    </BarChart>
  </ResponsiveContainer>
</div>
`;

exports[`FreeBarChart when data should allow personalizing the options 1`] = `
<div
  className="BarChart custom-class"
>
  <ResponsiveContainer
    debounce={0}
    height={320}
    width="100%"
  >
    <BarChart
      barCategoryGap="10%"
      barGap={4}
      data={
        Array [
          Object {
            "label": "col1",
            "lastProp": 2,
            "otherProp": 5,
            "results": 3,
          },
          Object {
            "label": "col2",
            "lastProp": 3,
            "otherProp": 2,
            "results": 4,
          },
        ]
      }
      height={300}
      layout="vertical"
      margin={
        Object {
          "bottom": 5,
          "left": 5,
          "right": 5,
          "top": 5,
        }
      }
      reverseStackOrder={false}
      stackOffset="none"
    >
      <XAxis
        allowDataOverflow={false}
        allowDecimals={true}
        allowDuplicatedCategory={true}
        dataKey="otherProp"
        domain={
          Array [
            0,
            "auto",
          ]
        }
        height={30}
        hide={false}
        mirror={false}
        orientation="bottom"
        padding={
          Object {
            "left": 0,
            "right": 0,
          }
        }
        reversed={false}
        scale="auto"
        tickCount={5}
        type="number"
        width={0}
        xAxisId={0}
      />
      <YAxis
        allowDataOverflow={false}
        allowDecimals={true}
        allowDuplicatedCategory={true}
        dataKey="label"
        domain={
          Array [
            0,
            "auto",
          ]
        }
        height={0}
        hide={false}
        mirror={false}
        orientation="left"
        padding={
          Object {
            "bottom": 0,
            "top": 0,
          }
        }
        reversed={false}
        scale="auto"
        tickCount={5}
        type="category"
        width={200}
        yAxisId={0}
      />
      <CartesianGrid
        fill="none"
        horizontal={true}
        horizontalFill={Array []}
        horizontalPoints={Array []}
        stroke="#ccc"
        strokeDasharray="2 4"
        vertical={true}
        verticalFill={Array []}
        verticalPoints={Array []}
      />
      <Tooltip
        active={false}
        animationDuration={400}
        animationEasing="ease"
        content={
          <CustomTooltip
            active={false}
          />
        }
        contentStyle={Object {}}
        coordinate={
          Object {
            "x": 0,
            "y": 0,
          }
        }
        cursor={true}
        cursorStyle={Object {}}
        filterNull={true}
        isAnimationActive={true}
        itemSorter={[Function]}
        itemStyle={Object {}}
        labelStyle={Object {}}
        offset={10}
        separator=" : "
        useTranslate3d={false}
        viewBox={
          Object {
            "x1": 0,
            "x2": 0,
            "y1": 0,
            "y2": 0,
          }
        }
        wrapperStyle={Object {}}
      />
      <Bar
        animationBegin={0}
        animationDuration={400}
        animationEasing="ease"
        data={Array []}
        dataKey="otherProp"
        fill="rgb(30, 28, 138)"
        hide={false}
        isAnimationActive={true}
        layout="vertical"
        legendType="rect"
        minPointSize={0}
        onAnimationEnd={[Function]}
        onAnimationStart={[Function]}
        stackId="unique"
        xAxisId={0}
        yAxisId={0}
      >
        <Cell
          cursor="pointer"
          fill="#ff0"
          key="cell-0"
        />
        <Cell
          cursor="pointer"
          fill="#00f"
          key="cell-1"
        />
      </Bar>
    </BarChart>
  </ResponsiveContainer>
</div>
`;

exports[`FreeBarChart when data should render basic data with correct defaults 1`] = `
<div
  className="BarChart "
>
  <ResponsiveContainer
    debounce={0}
    height={320}
    width="100%"
  >
    <BarChart
      barCategoryGap="10%"
      barGap={4}
      data={
        Array [
          Object {
            "label": "col1",
            "lastProp": 2,
            "otherProp": 5,
            "results": 3,
          },
          Object {
            "label": "col2",
            "lastProp": 3,
            "otherProp": 2,
            "results": 4,
          },
        ]
      }
      height={300}
      layout="horizontal"
      margin={
        Object {
          "bottom": 5,
          "left": 5,
          "right": 5,
          "top": 5,
        }
      }
      reverseStackOrder={false}
      stackOffset="none"
    >
      <XAxis
        allowDataOverflow={false}
        allowDecimals={true}
        allowDuplicatedCategory={true}
        dataKey="label"
        domain={
          Array [
            0,
            "auto",
          ]
        }
        height={30}
        hide={false}
        label=""
        mirror={false}
        orientation="bottom"
        padding={
          Object {
            "left": 0,
            "right": 0,
          }
        }
        reversed={false}
        scale="auto"
        tickCount={5}
        type="category"
        width={0}
        xAxisId={0}
      />
      <YAxis
        allowDataOverflow={false}
        allowDecimals={true}
        allowDuplicatedCategory={true}
        domain={
          Array [
            0,
            "auto",
          ]
        }
        height={0}
        hide={false}
        mirror={false}
        orientation="left"
        padding={
          Object {
            "bottom": 0,
            "top": 0,
          }
        }
        reversed={false}
        scale="auto"
        tickCount={5}
        type="number"
        width={60}
        yAxisId={0}
      />
      <CartesianGrid
        fill="none"
        horizontal={true}
        horizontalFill={Array []}
        horizontalPoints={Array []}
        stroke="#ccc"
        strokeDasharray="2 4"
        vertical={true}
        verticalFill={Array []}
        verticalPoints={Array []}
      />
      <Tooltip
        active={false}
        animationDuration={400}
        animationEasing="ease"
        content={
          <CustomTooltip
            active={false}
          />
        }
        contentStyle={Object {}}
        coordinate={
          Object {
            "x": 0,
            "y": 0,
          }
        }
        cursor={true}
        cursorStyle={Object {}}
        filterNull={true}
        isAnimationActive={true}
        itemSorter={[Function]}
        itemStyle={Object {}}
        labelStyle={Object {}}
        offset={10}
        separator=" : "
        useTranslate3d={false}
        viewBox={
          Object {
            "x1": 0,
            "x2": 0,
            "y1": 0,
            "y2": 0,
          }
        }
        wrapperStyle={Object {}}
      />
      <Bar
        animationBegin={0}
        animationDuration={400}
        animationEasing="ease"
        data={Array []}
        dataKey="results"
        fill="rgb(30, 28, 138)"
        hide={false}
        isAnimationActive={true}
        layout="vertical"
        legendType="rect"
        minPointSize={0}
        onAnimationEnd={[Function]}
        onAnimationStart={[Function]}
        stackId="unique"
        xAxisId={0}
        yAxisId={0}
      >
        <Cell
          cursor="pointer"
          fill="#1e1c8a"
          key="cell-0"
        />
        <Cell
          cursor="pointer"
          fill="#1e1c8a"
          key="cell-1"
        />
      </Bar>
    </BarChart>
  </ResponsiveContainer>
</div>
`;
